File: HeeeNoScreenShotView-demo/HeeeNoScreenShotView-demo-OC/HeeeNoScreenShotView-demo-OC/HeeeNoScreenShotView.m
//
//  HeeeNoScreenShotView.m
//  HeeeNoScreenShotView-demo-OC
//
//  Created by Heee on 2021/5/18.
//  注意：这个方法仅适用于iOS13及以后的系统

#import "HeeeNoScreenShotView.h"

@interface HeeeNoScreenShotView ()
@property (nonatomic,strong) UITextField *textField;
@property (nonatomic,strong) UIView *clearView;

@end

@implementation HeeeNoScreenShotView
- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (instancetype)init {
  self = [super init];
  if (self) {
    [self setupUI];
  }
  return self;
}

- (instancetype)initWithFrame:(CGRect)frame {
  self = [super initWithFrame:frame];
  if (self) {
    [self setupUI];
  }
  return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
  self = [super initWithCoder:coder];
  if (self) {
    [self setupUI];
  }
  return self;
}

- (void)layoutSubviews {
  [super layoutSubviews];
  self.textField.frame = self.bounds;
  self.clearView.frame = self.bounds;
}

- (void)setupUI {
  [self addSubview:self.textField];
  self.textField.subviews.firstObject.userInteractionEnabled = YES;
  [self.textField.subviews.firstObject addSubview:self.clearView];
  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
}

- (void)keyboardWillShow:(NSNotification *)noti {
    if (self.textField.isFirstResponder) {
        [self.textField resignFirstResponder];
        self.textField.subviews.firstObject.userInteractionEnabled = YES;
    }
}

- (void)addSubview:(UIView *)view {
  [super addSubview:view];
  if (self.textField != view) {
    [self.clearView addSubview:view];
  }
}

- (UITextField *)textField {
  if (!_textField) {
    _textField = [[UITextField alloc] init];
    _textField.secureTextEntry = YES;
  }
  
  return _textField;
}

- (UIView *)clearView {
  if (!_clearView) {
    _clearView = [[UIView alloc] init];
    _clearView.backgroundColor = [UIColor clearColor];
  }
  
  return _clearView;
}

@end

File: HeeeNoScreenShotView-demo/HeeeNoScreenShotView-demo-swift/HeeeNoScreenShotView-demo-swift/HeeeNoScreenShotView.swift
//
//  HeeeNoScreenShotView.swift
//  HeeeNoScreenShotView-demo-swift
//
//  Created by Heee on 2021/5/18.
//  注意：这个方法仅适用于iOS13及以后的系统

import UIKit

class HeeeNoScreenShotView: UIView {
  var textField : UITextField = {
    let tf = UITextField()
    tf.isSecureTextEntry = true
    return tf
  }()
  
  var clearView : UIView = {
    return UIView()
  }()
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
  init() {
    super.init(frame: .zero)
    self.setupUI()
  }
  
  override init(frame: CGRect) {
    super.init(frame: frame)
    self.setupUI()
  }
  
  required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
  }
  
  func setupUI() {
    self.addSubview(textField)
    if textField.subviews.first != nil {
      textField.subviews.first!.isUserInteractionEnabled = true
      textField.subviews.first!.addSubview(clearView)
    }
    
    NotificationCenter.default.removeObserver(self)
    NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow), name: UIResponder.keyboardWillShowNotification, object: nil)
  }
  
  @objc func keyboardWillShow() {
      if (textField.isFirstResponder) {
          textField.resignFirstResponder()
          if textField.subviews.first != nil {
            textField.subviews.first!.isUserInteractionEnabled = true
          }
      }
  }
    
  override func layoutSubviews() {
    super.layoutSubviews()
    textField.frame = self.bounds
    clearView.frame = self.bounds
      
    if (textField.superview != self) {
        self.addSubview(textField)
    }
  }
  
  override func addSubview(_ view: UIView) {
    super.addSubview(view)
    if textField != view {
      clearView.addSubview(view)
    }
  }
}
